doc/aaa/TACACS+ Design.md

```diff
@@ -0,0 +1,537 @@
+ - Support config multiple TACACS server.
+ - First server in the list will be primary server.
+ - When a server not accessible, will try next server as backup.
+ - When all server not accessible from SONiC, use native failover solution.
+
+# 2 Configuration and Management Requirements
+## 2.1 SONiC CLI
+ - Enable/Disable TACACS Authorization/Accounting command
+```
+	config aaa authorization {local | tacacs+}
+    config aaa accounting {local | tacacs+ | disable}
+```
+
+ - Counter command
+```
+    show tacacs counter
+    sonic-clear tacacscounters
+```
+
+## 2.2 Config DB
+ - TACACS AAA are fully configurable by config DB.
+
+## 2.3 Counter
+ - Support  AAA counter, this will be low priority:
+```
+    show tacacs counter
+    
+        server1: 10.1.1.45
+        Messages sent: 24
+        Messages received: 20
+        Requests accepted: 14
+        Requests rejected: 8
+        Requests timeout: 2
+        Requests retransmitted: 1
+        Bad responses: 1
+```
+
+## 2.4 Syslog
+ - Generate syslog when Authentication/Authorization/Accounting.
+ - When remote TACACS server not accessible from SONiC, use syslog for accounting.
+
+
+# 3 Limitation
+## 3.1 Command size
+ - TACACS protocol limitation: command + parameter size should smaller than 240 byte. The longer than 240 bytes parts will be drop.
+ 	- This limitation is a protocol level, all TACACS implementation have this limitation, include CISCO, ARISTA and Cumulus.
+ 	- Both Authorization and Accounting have this limitation.
+ 	- When user user a command longer than 240 bytes, only commands within 240 bytes will send to TACACS server. which means Accounting may lost some user input. and Authorization check can only partly check user input.
+
+## 3.2 Server count
+ - Max TACACS server  count was hardcoded, default count is 8.
+
+## 3.3 Local authorization
+ - Operation system limitation: SONiC based on Linux system, so permission to execute local command are managed by Linux file permission control. This means when enable both TACACS+ authorization and local authorization, local authorization will always happen after TACACS+ authorization.
+
+## 3.4 Docker support
+ - Any command run inside docker container will not covered by Authorization and Accounting.
+ - Docker exec command will be covered by Authorization and Accounting. 
+ - Administrator may setup TACACS+ rules to block docker exec command for RO user:
+   - user can start a interactive shell on the docker container, then run command inside container to evade TACACS+ authorization and accounting.
+
+## 3.5 Recursive commands
+  - Many linux command allow user start a harmless process and run another command from it,  administrator may setup TACACS+ rules from server side to block user from:
+    - Run another shell.
+    - Run interpreter, for example python.
+    - Run loader, for example /lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
+    - Run find/VI command which can run other commands inside it.
```

Comment:
>find/VI [](http://example.com/codeflow?start=10&length=7)

I can understand vim. `find` is similiar to other normal command and we can use regex to limit its power, right?

Reply:
Yes, we can use regex to limit user run 'find' with -exec parameter.

Reply:
Verified with POC code, we can block -exec parameter.